perfkitbenchmarker/providers/ibmcloud/ibmcloud_virtual_machine.py

```diff
@@ -0,0 +1,599 @@
+_WAIT_TIME_UBUNTU = 600
+_WAIT_TIME_DEBIAN = 120
+
+
+class IbmCloudVirtualMachine(virtual_machine.BaseVirtualMachine):
+  """Object representing a IBM Cloud Virtual Machine"""
+
+  CLOUD = providers.IBMCLOUD
+  IMAGE_NAME_PREFIX = None
+
+  _lock = threading.Lock()
+  command_works = False
+  ibmcloud_apikey = None
+  ibmcloud_account_id = None
+  validated_resources_set = set()
+  validated_subnets = 0
+
+  def __init__(self, vm_spec: virtual_machine.BaseVmSpec):
+    """Initialize a IBM Cloud virtual machine.
+    Args:
+      vm_spec: virtual_machine.BaseVirtualMachineSpec object of the vm.
+    """
+    super(IbmCloudVirtualMachine, self).__init__(vm_spec)
+    self.user_name = FLAGS.ibmcloud_image_username
+    self.boot_volume_size = FLAGS.ibmcloud_boot_volume_size
+    self.boot_volume_iops = FLAGS.ibmcloud_boot_volume_iops
+    self.volume_iops = _DEFAULT_VOLUME_IOPS
+    if FLAGS.ibmcloud_volume_iops:
+      self.volume_iops = FLAGS.ibmcloud_volume_iops
+    self.volume_profile = FLAGS.ibmcloud_volume_profile
+    self.image = FLAGS.image
+    self.os_data = None
+    self.user_data = None
+    self.vmid = None
+    self.vm_deleted = False
+    self.vm_started = False
+    self.instance_start_failed = True
+    self.instance_stop_failed = False
+    self.profile = FLAGS.machine_type
+    self.prefix = FLAGS.ibmcloud_prefix
+    self.zone = 'us-south-1'  # default
+    self.fip_address = None
+    self.fip_id = None
+    self.ssh_pub_keyfile = None
+    self.network = None
+    self.subnet = FLAGS.ibmcloud_subnet
+    self.subnets = {}
+    self.vpcid = FLAGS.ibmcloud_vpcid
+    self.key = FLAGS.ibmcloud_pub_keyid
+    self.sgid = None
+    self.boot_encryption_key = None
+    self.data_encryption_key = None
+    self.extra_vdisks_created = False
+    self.device_paths_detected = set()
+
+  def _CreateRiasKey(self):
+    """creates a ibmcloud key from the generated ssh key """
+    logging.info('Creating rias key')
+    with open(vm_util.GetPublicKeyPath(), 'r') as keyfile:
+      pubkey = keyfile.read()
+    logging.info('ssh private key file: %s, public key file: %s', \
+                 vm_util.GetPrivateKeyPath(), vm_util.GetPublicKeyPath())
+    cmd = ibm.IbmAPICommand(self)
+    cmd.flags['name'] = self.prefix + str(flags.FLAGS.run_uri) + 'key'
+    cmd.flags['pubkey'] = pubkey
+    return cmd.CreateKey()
+
+  def _CheckImage(self):
+    """Verifies we have an imageid to use """
+    cmd = ibm.IbmAPICommand(self)
+    cmd.flags['image_name'] = self.image or self._GetDefaultImageName()
+    logging.info('Looking up image: %s', cmd.flags['image_name'])
+    self.imageid = cmd.GetImageId()
+    if self.imageid is None:
+      logging.info('Failed to find valid image id')
+      os._exit(1)
+    else:
+      logging.info('Image id found: %s', self.imageid)
+
+  @classmethod
+  def _GetDefaultImageName(cls):
+    """Returns the default image name prefx """
+    return cls.IMAGE_NAME_PREFIX
+
+  def _SetupMzone(self):
```

Comment:
What's an "M" zone?

Reply:
Mzone is just an internal term we use but may not be clear, renaming to _SetupResources() instead.